apps/CA.pl.in

```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
         copy_pemfile($FILE,"${CATOP}/$CACERT", "CERTIFICATE");
     } else {
         print "Making CA certificate ...\n";
-        $RET = run("$REQ -new -keyout"
+        $RET = run("$REQ $OPENSSL_CONFIG -new -keyout"
                 . " ${CATOP}/private/$CAKEY"
                 . " -out ${CATOP}/$CAREQ");
-        $RET = run("$CA -create_serial"
+        $RET = run("$CA $OPENSSL_CONFIG -create_serial"
                 . " -out ${CATOP}/$CACERT $CADAYS -batch"
                 . " -keyfile ${CATOP}/private/$CAKEY -selfsign"
                 . " -extensions v3_ca"
                 . " -infiles ${CATOP}/$CAREQ") if $RET == 0;
         print "CA certificate is in ${CATOP}/$CACERT\n" if $RET == 0;
     }
-} elsif ($WHAT eq '-pkcs12' ) {
-    my $cname = $ARGV[1];
-    $cname = "My Certificate" unless defined $cname;
+    $RET = run("$X509 -in ${CATOP}/$CACERT -inform pem"
+            . " -out ${CATOP}/$CACERTDER -outform der") if $RET == 0;
+    print "DER format version is in ${CATOP}/$CACERTDER\n" if $RET == 0;
+}
+
+# Create a PKCS#12 file from the signed cert and key
+sub pkcs12
+{
+    my $cname = shift;
+    $cname = "$CNAMETXT" unless $cname;
     $RET = run("$PKCS12 -in $NEWCERT -inkey $NEWKEY"
             . " -certfile ${CATOP}/$CACERT"
-            . " -out $NEWP12"
+            . " -out $NEWP12 $PASSOPT"
             . " -export -name \"$cname\"");
-    print "PKCS #12 file is in $NEWP12\n" if $RET == 0;
-} elsif ($WHAT eq '-xsign' ) {
-    $RET = run("$CA -policy policy_anything -infiles $NEWREQ");
-} elsif ($WHAT eq '-sign' ) {
-    $RET = run("$CA -policy policy_anything -out $NEWCERT -infiles $NEWREQ");
+    print "PKCS#12 file is in $NEWP12\n" if $RET == 0;
+}
+
+# Sign a certificate based on a certificate request
+sub xsign
+{
+    my ($days,$day) = @_;
+    $days = "" unless $days;
+    if ($days eq '-days' and $day) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```

Comment:
It's quite unclear that you're actually getting `@ARGV` passed through `@_` and that this code is actually parsing command line arguments.  You *could* do the latter in a more readable form:

``` perl
    my $days = "";

    while (scalar @_ > 0) {
        if ($_[0] eq '-days') {
            die "-days must have a value\n" unless scalar @_ > 1;
            $days = join(" ", shift @_, shift @_);
        } else {
            last;
        }
    }
    die "extra arguments to xsign\n" unless scalar @_ == 0;
```